### PR TITLE
Add BitcoinDiamond (BCD) support

### PR DESCRIPTION
**Pull request by atomlab**
This PR adds support for the [BitcoinDiamond](https://github.com/eveybcd/BitcoinDiamond) cryptocurrency into Blockbook.

Coin information:

- https://www.bitcoindiamond.org/
- https://github.com/eveybcd/BitcoinDiamond
- https://coinmarketcap.com/currencies/bitcoin-diamond

I have setup a Blockbook server accessible here https://bitcoindiamond.atomicnodes.com/status

### CODE REVIEW

**Comment by martinboehm** [comment]
Hi, I tried to build the package and got this error when building backend:

```
dpkg-shlibdeps: error: cannot find library libX11-xcb.so.1 needed by debian/backend-bdiamond/opt/coins/nodes/bdiamond/bin/bitcoindiamond-qt (ELF format: 'elf64-x86-64' abi: '0201003e00000000'; RPATH: '')
```

**Comment by atomlab** [comment]
Thank you for you review! I have updated [Dockerfile](https://github.com/Atomicwallet/blockbook/blob/c7e88c5e522921262bd392f23975582b006a3c30/build/docker/bin/Dockerfile#L9) and add libx11-xcb1 dependency. This error should be fixed. 

But now I'm see other error
```
FAIL
Makefile:28: recipe for target 'test' failed
make[1]: *** [test] Error 1
make[1]: Leaving directory '/build/build/pkg-defs/blockbook'
dh_auto_test: make -j1 test returned exit code 2
debian/rules:6: recipe for target 'build' failed
make: *** [build] Error 2
dpkg-buildpackage: error: debian/rules build gave error exit status 2
Makefile:34: recipe for target 'deb-bdiamond' failed
make: *** [deb-bdiamond] Error 2
```

I don't understand what happens bu I will try fix it.

**Comment by liyaojian** [comment]
@atomlab hi, I try [your repo](https://github.com/Atomicwallet/blockbook/tree/bdiamond) for `/data/blockbook/blockbook -sync -blockchaincfg=/data/blockbook/build/blockchaincfg.json -internal=:19046 -public=:19146 -datadir=/data/block_data/bcd/ -logtostderr`

sync data pretty slowly and how long did you sync completely ?

**Comment by atomlab** [comment]
> @atomlab hi, I try [your repo](https://github.com/Atomicwallet/blockbook/tree/bdiamond) for `/data/blockbook/blockbook -sync -blockchaincfg=/data/blockbook/build/blockchaincfg.json -internal=:19046 -public=:19146 -datadir=/data/block_data/bcd/ -logtostderr`
> 
> sync data pretty slowly and how long did you sync completely ? 

Hi! Yeah,  I know, sync takes about 5 days. Unfortunately I don't know how to improve it.

**Comment by liyaojian** [comment]
but 15 days passed , It's still not over. maybe some problem with me ? may i have your operating instructions. thanks!

**Comment by liyaojian** [comment]
bcd node and blockbook are in one server(4cpu/12GB/512G HDD). 
run `/data/blockbook/blockbook -sync -blockchaincfg=/data/blockbook/build/blockchaincfg.json -internal=:19046 -public=:19146 -datadir=/data/block_data/bcd/ -logtostderr` in supervisor.

above. there are some problem or improve ?

**Comment by atomlab** [comment]
In my case blockbook works on server AMD Ryzen 7 3700X/8CPU/25GB RAM/1TB NVME SSD. 
During sync the blockbook takes about 15GB RAM. 
Blockbook sync takes about 5 days. 
After sync blockbook takes about 5GB. 
Blockbook db size 128GB.
My bitcoin-diamond node runs on the same host.

I run blockbook with params
```
blockbook -blockchaincfg=/config/blockchaincfg.json -datadir=/data/db -sync -internal=:9035 -public=:9135 -explorer= -log_dir=/logs -dbcache=0 -workers=1 -logtostderr -debug
```

I think HDD not best choice for rocksdb (blockbook use it).

**Comment by liyaojian** [comment]
Thank you very much for your information.

**Review by IKE471 — APPROVED** [review]

**Review by IKE471 — APPROVED** [review]

**Comment by atomlab** [comment]
I have an error with build. I'm working on this problem.

```
GetChainParams case chain: mainnet
--- FAIL: TestGetAddrDesc (0.00s)
    --- FAIL: TestGetAddrDesc/bcd-2 (0.00s)
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0

goroutine 62 [running]:
testing.tRunner.func1.1(0x907e20, 0xc0000247e0)
	/opt/go/src/testing/testing.go:940 +0x2f5
testing.tRunner.func1(0xc000146900)
	/opt/go/src/testing/testing.go:943 +0x3f9
panic(0x907e20, 0xc0000247e0)
	/opt/go/src/runtime/panic.go:969 +0x166
github.com/trezor/blockbook/bchain/coins/bcd.TestGetAddrDesc.func1(0xc000146900)
	/go/src/github.com/trezor/blockbook/bchain/coins/bcd/bdiamondparser_test.go:126 +0x480
testing.tRunner(0xc000146900, 0xc0000f30d0)
	/opt/go/src/testing/testing.go:991 +0xdc
created by testing.(*T).Run
	/opt/go/src/testing/testing.go:1042 +0x357
FAIL	github.com/trezor/blockbook/bchain/coins/bcd	0.075s
```

**Comment by atomlab** [comment]
all errors fixed
```
$ make all-bdiamond
$ find ./build/ -name "*.deb"
./build/backend-bdiamond_1.3.0-satoshilabs-1_amd64.deb
./build/blockbook-bdiamond_0.3.4_amd64.deb
```

**Comment by magmel48** [comment]
is there any update on this?

File: bchain/coins/blockchain.go [hunk 1]
```diff
@@ -70,6 +71,7 @@ func init() {
 	BlockChainFactories["Ethereum Testnet Ropsten"] = eth.NewEthereumRPC
 	BlockChainFactories["Bcash"] = bch.NewBCashRPC
 	BlockChainFactories["Bcash Testnet"] = bch.NewBCashRPC
+  BlockChainFactories["Bdiamond"] = bcd.NewBdiamondRPC
```
Comment (magmel48) [inline bchain/coins/blockchain.go:74]: I think here is a problem with indentation

**Comment by martinboehm** [comment]
Hi. Sorry for the late response. 

Could you please do the following before we can merge the PR:
1) fix the problem with bitcoindiamond-qt by removing it from the package, not by adding a library to the Dockerfile. See how it is done for example in `configs/coins/bgold.json`
2) use ports from available mainnet range (for example 9068)
3) fix the indentation issue

**Comment by smithy9901** [comment]
All errors fixed

**Comment by atomlab** [comment]
@martinboehm Hi. Could you check us PR again please?

**Comment by martinboehm** [comment]
@atomlab Hi, when I try to build it, I get error:

**Comment by atomlab** [comment]
@martinboehm Sorry. My mistake. We have fixed this.

**Comment by martinboehm** [comment]
@atomlab Hi, I was able to build the packages now, however when I try to start the backend, I get the following error:
```
Aug 12 22:55:08 blockbook-dev systemd[6744]: backend-bdiamond.service: Failed to execute command: No such file or directory
Aug 12 22:55:08 blockbook-dev systemd[6744]: backend-bdiamond.service: Failed at step EXEC spawning /opt/coins/nodes/bdiamond/bin/bdiamondd: No such file or directory
```
and indeed, there is no such file in the directory:
```
$ ll /opt/coins/nodes/bdiamond/bin/                                                                                                                                                                                                                             [±Atomicwallet-bdiamond ✓]
Permissions Size User Date Modified Name
.rwxr-xr-x  3.0M root 12 Aug 22:30  bitcoindiamond-cli
.rwxr-xr-x  4.1M root 12 Aug 22:30  bitcoindiamond-tx
.rwxr-xr-x   11M root 12 Aug 22:30  bitcoindiamondd
```

**Comment by smithy9901** [comment]
@martinboehm Hi, error corrected, assembly, and launch tested.

**Comment by martinboehm** [comment]
Hi, sorry, the PR is obsolete now, I have to close it. If you want to add your coin to Blockbook, please open a new PR.

**Pull request closed without merging**